### DISCUSSION TITLE
How to access API endpoints using an app installation token?

### DISCUSSION THREAD

**Discussion by satwikkansal** [community/community#24502] [Apps, API and Webhooks]
<p>Hi, I’m developing a GitHub app in which I need to create repositories on the installed user’s Github. I have been following the official docs <a href="https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#accessing-api-endpoints-as-an-installation" rel="nofollow noopener">Authenticating with GitHub Apps</a> but it doesn’t seem to work.</p>
<p>I have successfully obtained the installation access token, but when calling the <a href="https://developer.github.com/v3/repos/#create-a-repository-for-the-authenticated-user" rel="nofollow noopener"> Create a repository for the authenticated user</a> API using the installation access token, I get the 403 error with the following message,</p>
<p><code>{'message': 'Resource not accessible by integration', 'documentation_url': 'https://developer.github.com/v3/repos/#create'}</code></p>
<p>For reference, my app has following permissions,</p>
<p><code>{'token': 'SOME_INSTALLATION_TOKEN', 'expires_at': '2020-07-13T12:38:17Z', 'permissions': {'administration': 'write', 'contents': 'write', 'metadata': 'read', 'pages': 'write'}, 'repository_selection': 'all'}</code></p>
<p>I’m not able to figure out what I’m doing wrong or how else to proceed.</p>

**Comment by francisfuzz** [comment]
<p>:wave: hello there <a class="mention" href="/satwikkansal">@satwikkansal</a>, and welcome to the GitHub Support Community!</p>
<aside class="quote no-group" data-username="satwikkansal" data-post="1" data-topic="122510">
<div class="title">
<div class="quote-controls"></div>
 satwikkansal:</div>
<blockquote>
<p>I have successfully obtained the installation access token, but when calling the <a href="https://developer.github.com/v3/repos/#create-a-repository-for-the-authenticated-user"> Create a repository for the authenticated user</a> API using the installation access token, I get the 403 error with the following message,</p>
<p><code>{'message': 'Resource not accessible by integration', 'documentation_url': 'https://developer.github.com/v3/repos/#create'}</code></p>
</blockquote>
</aside>
<p>As a next step for troubleshooting this further, could you please send us the full request-response pair of a <a href="http://curl.haxx.se/"><code>curl -v</code> request</a> that demonstrates what you’re seeing? That should help us investigate further.</p>
<p>Please make sure you mask any sensitive information like OAuth tokens and Authorization headers in the output of the curl command.</p>

**Comment by satwikkansal** [comment]
<p>Hi <a class="mention" href="/francisfuzz">@francisfuzz</a></p>
<p>Here’s the curl</p>
<pre><code class="lang-auto">curl --location --request POST 'https://api.github.com/user/repos' \
--header 'accept: application/vnd.github.machine-man-preview+json' \
--header 'authorization: token INSTALLATION_ACCESS_TOKEN_HERE' \
--header 'Content-Type: application/json' \
--data-raw '{
        "name": "test_repo",
        "description": "This is your first repository",
        "homepage": "https://github.com",
        "private": false,
        "has_issues": true,
        "has_projects": true,
        "has_wiki": true
    }'
</code></pre>

**Comment by francisfuzz** [comment]
<p><a class="mention" href="/satwikkansal">@satwikkansal</a> thanks for sharing that example <code>curl</code> request. The endpoint for <a href="https://docs.github.com/en/rest/reference/repos#create-a-repository-for-the-authenticated-user">creating a repository for the authenticated user</a> is not enabled for GitHub Apps.</p>
<p>Here are some ways to determine which endpoints are enabled for GitHub Apps</p>
<p>The first way is to look at the endpoint’s documentation’s <code>Notes</code> section. For example, the endpoint for <a href="https://docs.github.com/en/rest/reference/repos#list-repositories-for-a-user">listing public repositories for a specified user</a> is one such endpoint.</p>
<p>The second way is to consult the <a href="https://docs.github.com/en/rest/reference/permissions-required-for-github-apps"><code>Permissions required for GitHub Apps</code></a> article. A permission is associated with a collection of endpoints. If an endpoint you’re looking for isn’t listed in any of those permissions, then it’s not enabled for GitHub Apps.</p>
<p>The third way is to consult <a href="https://github.com/octokit/routes"><code>octokit/routes</code></a>, a project that contains a machine-readable, always up-to-date GitHub REST API route specifications. Any usage or troubleshooting questions are best fielded in <a href="https://github.com/octokit/routes/issues/new">a form of a new issue in that repository</a>. In relation to the endpoint you’re using, this library leverages the <a href="https://swagger.io/specification/">OpenAPI Specification</a> and encapsulates each endpoint into its own JSON file. For example, the endpoint for <a href="https://github.com/octokit/routes/blob/da86230b26f5f53809d837d44888479779be0661/openapi/api.github.com/operations/repos/create-for-authenticated-user.json">creating a repository for the authenticated user</a> is defined and also includes an <a href="https://github.com/octokit/routes/blob/da86230b26f5f53809d837d44888479779be0661/openapi/api.github.com/operations/repos/create-for-authenticated-user.json#L267-L283"><code>x-github</code> field</a> specifying some helpful metadata. In particular, the <code>enabledForApps</code> field’s value is <code>false</code>, indicating that it’s not an endpoint a GitHub App can access:</p>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/octokit/routes/blob/da86230b26f5f53809d837d44888479779be0661/openapi/api.github.com/operations/repos/create-for-authenticated-user.json#L269" target="_blank">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/octokit/routes/blob/da86230b26f5f53809d837d44888479779be0661/openapi/api.github.com/operations/repos/create-for-authenticated-user.json#L269" target="_blank">octokit/routes/blob/da86230b26f5f53809d837d44888479779be0661/openapi/api.github.com/operations/repos/create-for-authenticated-user.json#L269</a></h4>
<pre class="onebox"><code class="lang-json"><ol class="start lines" start="259" style="counter-reset: li-counter 258 ;">
<li>    "lang": "JS",</li>
<li>    "source": "octokit.repos.createForAuthenticatedUser({\n  name: 'name'\n})"</li>
<li>  },</li>
<li>  {</li>
<li>    "lang": "Ruby",</li>
<li>    "source": "octokit.create_for_authenticated_user(\n  'name'\n)"</li>
<li>  }</li>
<li>],</li>
<li>"x-github": {</li>
<li>  "legacy": false,</li>
<li class="selected">  "enabledForApps": false,</li>
<li>  "githubCloudOnly": false,</li>
<li>  "previews": [</li>
<li>    {</li>
<li>      "name": "nebula",</li>
<li>      "required": false,</li>
<li>      "note": "You can set the visibility of a repository using the new `visibility` parameter in the [Repositories API](https://developer.github.com/v3/repos/), and get a repository's visibility with a new response key. For more information, see the [blog post](https://developer.github.com/changes/2019-12-03-internal-visibility-changes/).\n\nTo access repository visibility during the preview period, you must provide a custom [media type](https://developer.github.com/v3/media) in the `Accept` header:\n```shell\napplication/vnd.github.nebula-preview+json\n```"</li>
<li>    },</li>
<li>    {</li>
<li>      "name": "baptiste",</li>
<li>      "required": false,</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Hope this helps! :bowing_man:</p>

**Comment by satwikkansal** [comment]
<p><a class="mention" href="/francisfuzz">@francisfuzz</a> thank you! I get it now.</p>
<p>Is there any other way a third party can create a repository in the user’s account?</p>
<p>I tried looking into the docs, but it seems like this can only be done with the help of personal access tokens. But that seems unlikely, I’m probably missing something but not sure what.</p>

**Comment by francisfuzz** [comment, marked as answer]
<aside class="quote no-group" data-username="satwikkansal" data-post="5" data-topic="122510">
<div class="title">
<div class="quote-controls"></div>
 satwikkansal:</div>
<blockquote>
<p>Is there any other way a third party can create a repository in the user’s account?</p>
</blockquote>
</aside>
<p><a class="mention" href="/satwikkansal">@satwikkansal</a> One approach is <a href="https://docs.github.com/en/developers/apps/creating-an-oauth-app">creating an OAuth app</a> that requests either the <code>public_repo</code> scope (ability to create only public repositories) or the <code>repo</code> scope (ability to create either public or private repositories).</p>
<p>Once that person <a href="https://docs.github.com/en/developers/apps/authorizing-oauth-apps#web-application-flow">authorizes your OAuth app through the web application flow</a> with either scope, then you can <a href="https://docs.github.com/en/developers/apps/authorizing-oauth-apps#3-use-the-access-token-to-access-the-api">use the access token to access the API</a> on their behalf (like <a href="https://docs.github.com/en/rest/reference/repos#create-a-repository-for-the-authenticated-user">creating a repository</a> on their account). Hope this helps!</p>

**Comment by satwikkansal** [comment]
<p><a class="mention" href="/francisfuzz">@francisfuzz</a> thanks again :slight_smile: That mostly answers my question.</p>
<p>I have just one last question, will be able to list an oauth app just like Github app on the marketplace, or is there any caveat I should be aware of?</p>

**Comment by francisfuzz** [comment]
<aside class="quote no-group" data-username="satwikkansal" data-post="7" data-topic="122510">
<div class="title">
<div class="quote-controls"></div>
 satwikkansal:</div>
<blockquote>
<p>I have just one last question, will be able to list an oauth app just like Github app on the marketplace, or is there any caveat I should be aware of?</p>
</blockquote>
</aside>
<p>Yes! It’s possible to <a href="https://docs.github.com/en/developers/github-marketplace/drafting-a-listing-for-your-app">draft a listing for either an OAuth app or a GitHub App</a> depending on what you’d like to add to the GitHub Marketplace.</p>
